templates/NDArray.kt

```diff
@@ -90,6 +90,53 @@ interface NDArray<T> {
      */
     fun toMutableList() = MutableList(size) { getGeneric(it) }
 
+    /**
+     * Determine whether every element of one NDArray is sufficiently close to the corresponding element
+     * of another array.  Two elements are considered close if
+     *
+     * abs(ele1 - ele2) < (atol + rtol * abs(ele1))
+     *
+     * @param other   the array to compare this one to
+     * @param rtol    the maximum relative (i.e. fractional) difference to allow between elements
+     * @param atol    the maximum absolute difference to allow between elements
+     */
+    fun <R> allClose(other: NDArray<R>, rtol: Double=1e-05, atol: Double=1e-08): Boolean {
```

Comment:
Adding `allClose` directly to the `NDArray` interface means it will be available to all `NDArray`s, including e.g. `NDArray<String>`. One of the goals of implementing it [as an extension function on Matrix](https://github.com/kyonifer/koma/blob/master/koma-core-api/common/src/koma/extensions/extensions_matrix_Double.kt#L285) was that it would only appear when a user had a supported numeric matrix type. For example, `Matrix<Double>.allClose` would compile but `Matrix<Complex>` wouldn't. With the implementation in this PR, `ndArrayOf("a","b").allClose(ndArrayOf("a","b"))` compiles and throws an exception at runtime and we therefore lose type safety.

I agree with removing the `Matrix` version of `allClose` (since NDArray is a super type), but I'd recommend keeping the implementation as a set of extension functions, optimally only implemented for floating point (since as you pointed out in the issue, these functions don't make sense for integral types). If the difficulty in doing so is the currently complicated codegen implementation, I can take over from here and work that in if you'd like.

Reply:
How about defining it as an extension function for type `<out Number>`?  That way you could still use it to compare between different numerical types, like I do in the test case.

Reply:
Unless we can guarantee that `getDouble()` will always work, I think boxing is unavoidable.  What about modifying the generic version to cast between numeric types, just like the specialized ones do?  For example, in `DefaultGenericNDArray.getDouble()` change

```kotlin
if (ele is Double)
    return ele
```

to

```kotlin
if (ele is Number)
    return ele.toDouble()
```

 Likewise, require that any future backends also support casting.  Then we can just call `getDouble()` and know it will be implemented in whatever way is most efficient given the storage of the particular array.

Reply:
I was thinking we could dispatch in a `when` block, similar to the `NDArray.invoke` block, that checked the reified type and called a specific one for each primitive, i.e. for a `NDArray<Int>` we could call `.getInt(..).toDouble()`. I haven't tried to see if it would break down though.

> if (ele is Number)
>    return ele.toDouble()

I think this route would work as well and might be cleaner. As long as we only expose the extension functions for NDArrays containing primitive types and we don't box in the specialized implementations, no objections here.

Reply:
Ok, I'll change it to work like that.

Reply:
Another option would be to not even have the extension function, and just put the whole implementation into the `allclose()` function.

Reply:
Inside the interface, you only get an opportunity to reify `<R>`, so you're still doing needlessly-expensive number conversion logic on the contents of the NDArray.

Reply:
I don't see why?  If `getDouble()` is declared to return a `Double`, on the JVM it will be implemented with a method whose return type is a primitive `double`.  It's up to each implementation of the interface to provide a value of that type.  In the case of an array of primitive doubles, no conversion or boxing should be needed.

Reply:
I think I misread much of the above. You guys were talking about boxing and I misread it as being about type conversion. My comment was trying to avoid doing _O(n)_ primitive casts from whatever the true underlying data type is to double, which seems like it might be important given

```kotlin
>>> Long.MAX_VALUE.toDouble() == Long.MAX_VALUE.toDouble() + 1
true
```